xdsl/backend/csl/print_csl.py

```diff
@@ -266,15 +385,43 @@ def descend(self) -> CslPrintContext:
             output=self.output,
             variables=self.variables.copy(),
             _counter=self._counter,
-            _prefix=self._prefix + "  ",
+            _prefix=self._prefix + self._INDENT,
         )
 
 
+def _get_layout_program(module: ModuleOp) -> tuple[csl.CslModuleOp, csl.CslModuleOp]:
+    """
+    Get the layout and program `csl.module`s from the top level `builtin.module`.
+
+    Makes sure there is exactly 1 layout and 1 program `csl.module`.
+
+    Returns layout first, then program.
+    """
+    ops_list = list(module.body.block.ops)
+    assert all(
+        isinstance(mod, csl.CslModuleOp) for mod in ops_list
+    ), "Expected all top level ops to be csl.module"
+    # We have asserted that this is true above
+    ops_list = cast(list[csl.CslModuleOp], ops_list)
+    assert len(ops_list) == 2, "Expected exactly two top level modules"
+    # This allows program and layout to be scpecified in any order
+    prog = next(
+        filter(lambda mod: mod.kind.data == csl.ModuleKind.PROGRAM, ops_list), None
+    )
+    layout = next(
+        filter(lambda mod: mod.kind.data == csl.ModuleKind.LAYOUT, ops_list), None
+    )
+    assert prog is not None, "Expected exactly 1 program module"
+    assert layout is not None, "Expected exactly 1 layout module"
+    return layout, prog
+
+
 def print_to_csl(prog: ModuleOp, output: IO[str]):
     """
     Takes a module op and prints it to the given output stream.
     """
     ctx = CslPrintContext(output)
-    for mod in prog.body.block.ops:
-        assert isinstance(mod, csl.CslModuleOp)
-        ctx.print_block(mod.body.block)
+    layout, program = _get_layout_program(prog)
+    ctx.print_block(program.body.block)
+    ctx.print(ctx.DIVIDER)
+    ctx.print_block(layout.body.block)
```

Comment:
I like the previous approach more of iterating over the modules, I feel this provides more flexibility for e.g. multiple programs or multiple layouts. This feels overly restrictive

Reply:
We have to print program first as that's when we collect symbols to be exported.

We can allow for multiple programs, but there still needs to be some logic to make sure they all get processed before the one layout.